pkg/proxy/iptables/proxier.go

```diff
@@ -843,6 +845,11 @@ func (proxier *Proxier) syncProxyRules() {
 			for _, extIP := range svcInfo.ExternalIPStrings() {
 				staleServices.Insert(extIP)
 			}
+			nodePort := svcInfo.NodePort()
+			if svcInfo.Protocol() == v1.ProtocolUDP && nodePort != 0 {
+				klog.V(2).Infof("Stale %s service NodePort %v -> %d", strings.ToLower(string(svcInfo.Protocol())), svcPortName, svcInfo.NodePort())
+				staleNodePorts.Insert(nodePort)
```

Comment:
It looks like this handles the case of the entire Service going away, but does it handle the case of a single endpoint going away / changing?

The "staleness" code is very confusing...

Reply:
yeah, I had to look it several times in the code to understand it, because my understanding was exactly what you are saying.However, the observation during debugging the problem is that the service ClusterIP was declared "stale" when the endpoint was updated but the service not, that is exactly the condition that we are looking for. The test added shows perfectly the behaviour.

I don´t know if this was on purpose or a side effect of how the endpoints/services tracker/map/cache works, but this is the kind of "staleness" that kube-proxy was  using always ... and this is exactly the place where we should delete the nodeport, that is also coherent with the rest of conntrack clearing

Reply:
aha, is already documented

https://github.com/kubernetes/kubernetes/blob/074a5177209c3605daf11cfe6f477fddce0d55c0/pkg/proxy/endpoints.go#L534-L536

Reply:
added an additional comment, but we must delete the nodeport conntrack entries when endpoints move from 0 to non zero, not when the service is added without endpoints.
This should be ready